File: java/src/main/java/com/epam/eng/pattern/facade/SmartHome.java
package com.epam.eng.pattern.facade;

class Light {
    void on() {
        System.out.println("Light is on");
    }

    void off() {
        System.out.println("Light is off");
    }
}

class Fan {
    void on() {
        System.out.println("Fan is on");
    }

    void off() {
        System.out.println("Fan is off");
    }
}

class AirCon {
    void on() {
        System.out.println("Air conditioner is on");
    }

    void off() {
        System.out.println("Air conditioner is off");
    }
}

class Home {
    private Light livingRoomLight;
    private Light kitchenLight;
    private Fan livingRoomFan;
    private Fan bedroomFan;
    private AirCon bedroomAC;

    public Home() {
        livingRoomLight = new Light();
        kitchenLight = new Light();
        livingRoomFan = new Fan();
        bedroomFan = new Fan();
        bedroomAC = new AirCon();
    }

    public void turnOnLights() {
        livingRoomLight.on();
        kitchenLight.on();
    }

    public void turnOffLights() {
        livingRoomLight.off();
        kitchenLight.off();
    }

    public void turnOnFan() {
        livingRoomFan.on();
        bedroomFan.on();
    }

    public void turnOffFan() {
        livingRoomFan.off();
        bedroomFan.off();
    }

    public void turnOnAc() {
        bedroomAC.on();
    }

    public void turnOffAc() {
        bedroomAC.off();
    }
}

public class SmartHome {
    public static void main(String[] args) {
        Home home = new Home();
        home.turnOnLights();
        home.turnOnFan();
        home.turnOnAc();

        home.turnOffLights();
        home.turnOffFan();
        home.turnOffAc();
    }
}

// todo: consider to turn On all devices by a single method. consider to turn Off all devices by a single method.

File: ts/facade/SmartHome.ts
class Light {
    on() {
      console.log("Light is on");
    }
  
    off() {
      console.log("Light is off");
    }
  }
  
  class Fan {
    on() {
      console.log("Fan is on");
    }
  
    off() {
      console.log("Fan is off");
    }
  }
  
  class AirCon {
    on() {
      console.log("Air conditioner is on");
    }
  
    off() {
      console.log("Air conditioner is off");
    }
  }
  
  class Home {
    private livingRoomLight: Light;
    private kitchenLight: Light;
    private livingRoomFan: Fan;
    private bedroomFan: Fan;
    private bedroomAC: AirCon;
  
    constructor() {
      this.livingRoomLight = new Light();
      this.kitchenLight = new Light();
      this.livingRoomFan = new Fan();
      this.bedroomFan = new Fan();
      this.bedroomAC = new AirCon();
    }
  
    turnOnLights() {
      this.livingRoomLight.on();
      this.kitchenLight.on();
    }
  
    turnOffLights() {
      this.livingRoomLight.off();
      this.kitchenLight.off();
    }
  
    turnOnFan() {
      this.livingRoomFan.on();
      this.bedroomFan.on();
    }
  
    turnOffFan() {
      this.livingRoomFan.off();
      this.bedroomFan.off();
    }
  
    turnOnAc() {
      this.bedroomAC.on();
    }
  
    turnOffAc() {
      this.bedroomAC.off();
    }
  }
  
  const smartHome = () => {
    const home = new Home();
    home.turnOnLights();
    home.turnOnFan();
    home.turnOnAc();
  
    home.turnOffLights();
    home.turnOffFan();
    home.turnOffAc();
  };
  
  smartHome();

  // todo: consider to turn On all devices by a single method. consider to turn Off all devices by a single method.
